clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedMessageSender.java

```diff
@@ -0,0 +1,304 @@
+        final Span span = startSpan(topic, tenantId, deviceId, context);
+
+        return send(topic, tenantId, deviceId, payload, headers, span);
+    }
+
+    private Future<Void> send(final String topic, final String tenantId, final String deviceId, final Buffer payload,
+            final List<KafkaHeader> headers, final Span span) {
+        final KafkaProducerRecord<String, Buffer> record = KafkaProducerRecord.create(topic, deviceId, payload);
+        final Promise<RecordMetadata> sendPromise = Promise.promise();
+
+        record.addHeaders(headers);
+        KafkaTracingHelper.injectSpanContext(tracer, record, span.context());
+        logProducerRecord(span, record);
+        getOrCreateProducer().send(record, sendPromise);
+
+        return sendPromise.future()
+                .recover(t -> {
+                    logError(span, topic, tenantId, deviceId, t);
+                    span.finish();
+                    return Future.failedFuture(new ServerErrorException(getErrorCode(t), t));
+                })
+                .map(recordMetadata -> {
+                    logRecordMetadata(span, deviceId, recordMetadata);
+                    span.finish();
+                    return null;
+                });
+    }
+
+    private void configureUniqueClientId() {
+        final UUID uuid = UUID.randomUUID();
+        final String clientId = config.get(ProducerConfig.CLIENT_ID_CONFIG);
+
+        if (clientId == null) {
+            config.put(ProducerConfig.CLIENT_ID_CONFIG, String.format("%s-%s", producerName, uuid));
+        } else {
+            config.put(ProducerConfig.CLIENT_ID_CONFIG, String.format("%s-%s-%s", clientId, producerName, uuid));
+        }
+    }
+
+    private KafkaProducer<String, Buffer> getOrCreateProducer() {
+        return producerFactory.getOrCreateProducer(producerName, config);
+    }
+
+    private List<KafkaHeader> encodePropertiesAsKafkaHeaders(final Map<String, Object> properties, final Span span) {
+        final List<KafkaHeader> headers = new ArrayList<>();
+
+        if (properties != null) {
+            properties.forEach((k, v) -> {
+                try {
+                    final Buffer headerValue = (v instanceof String)
+                            ? Buffer.buffer((String) v)
+                            : Buffer.buffer(Json.encode(v));
+
+                    headers.add(new KafkaHeaderImpl(k, headerValue));
+                } catch (final EncodeException e) {
+                    log.info("failed to serialize property with key [{}] to Kafka header", k);
+                    span.log("failed to create Kafka header from property: " + k);
+                }
+            });
+        }
+
+        return headers;
+    }
+
+    private Span startSpan(final String topic, final String tenantId, final String deviceId, final QoS qos,
+            final SpanContext context) {
+        final String referenceType = QoS.AT_MOST_ONCE.equals(qos) ? References.FOLLOWS_FROM : References.CHILD_OF;
```

Comment:
With my proposal above (providing separate `send` methods for both QoS levels), the QoS could be completely removed from the class and the reference type here determined depending on which `send` method is invoked.